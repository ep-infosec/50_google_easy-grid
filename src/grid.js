/**
 * Copyright 2017 Google LLC
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 * https://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import styled from 'styled-components';

 const GridDiv = styled('div')(({
	rowHeaders,
	colHeaders 
  }) => ({
	display: "grid",
	gridTemplateRows: rowHeaders || "initial",
	gridTemplateColumns: colHeaders || "initial"
  }))
  
   const GridCell = styled('div')(({ top, bottom, left, right }) => ({
	gridRowStart: top + 1,
	gridRowEnd: bottom + 2,
	gridColumnStart: left + 1,
	gridColumnEnd: right + 2,
	position: "relative"
  }))

const validateRows = (rows) => {
	const colsRegEx = /[\s]+/;
	const toValidate = rows.map(row => row.split(colsRegEx)).map(cols =>
		cols.reduce((acc, val) => {
			return val.length == 0 ? acc : acc.concat([val]);
		}, []));

	const colCounts = toValidate.map(row => row.length);
	if (colCounts.length == 1) {
		return toValidate;
	}
	const allSame = colCounts.slice(1).reduce((acc, val) => {
		return acc && val == colCounts[1];
	}, true);
	if (!allSame) {
		throw `Rows 1 - ${colCounts.length} should have the same number of ` +
			`columns.`;
	}
	if (colCounts[1] != 1 && colCounts[0] != colCounts[1] - 1) {
		throw `The column headers row has too many columns.`;
	}
	return toValidate;
};

const extractDimensions = (layout) => {
	const areas = {};
	for (let i = 0; i < layout.length; i++) {
		for (let j = 0; j < layout[i].length; j++) {
			if (layout[i][j] == '..') {
				continue;
			}
			const cellAreas = layout[i][j].split(',');
			cellAreas.forEach(area => {
				if (!areas[area]) {
					areas[area] = {
						top: i,
						left: j,
						bottom: i,
						right: j,
					};
				} else {
					areas[area] = {
						...areas[area],
						bottom: i,
						right: j,
					};
				}
			});
		}
	}
	return areas;
};

const GridComponentFactory = (...args) => {
	const layoutString = args[0][0];
	const rows = validateRows(layoutString.split('\n').reduce((acc, row) => {
		// Verify that the row is not empty.
		const match = row.match(/[\s]*/);
		if (row.length > 0 && match && match[0].length != match.input.length) {
			return acc.concat(row);
		}
		return acc;
	}, []));

	let rowHeaders = null;
	if (rows.length > 1) {
		// Remove the first row of layout data if this is a grid. If it's only
		// row headers, do not remove the first row of data.
		const rowHeadersToUse = rows[0].length > 1 ? rows.slice(1) : rows;
		rowHeaders = rowHeadersToUse.reduce((acc, row) => {
			return acc.concat([row[0]]);
		}, []).join(' ');
	}
	let colHeaders = null;
	if (rows[0].length > 1) {
		colHeaders = rows[0].join(' ');
	}
	let extractedDimensions = null;
	let sortedDimensionKeys = null;
	if (rows.length > 1 && rows[1].length > 1) {
		// Remove the column headers and row headers.
		const layout = rows.slice(1).map(row => row.slice(1));
		extractedDimensions = extractDimensions(layout);
		// The grid items are sorted alphabetically and matched against the
		// corresponding child element. Unless the child element contains a
		// gridItem property.
		sortedDimensionKeys = Object.keys(extractedDimensions).sort();
	}
	// Return a wrapper component that lays out its children using a generated
	// grid and grid items. Individual grid items are used to wrap the actual
	// child elements.
	const component = (props) => {
		let children = props.children;
		if (extractedDimensions) {
			const numberOfChildren = React.Children.count(props.children);
			if (numberOfChildren > sortedDimensionKeys.length) {
				throw `There are more children than defined areas for the grid.`;
			} else if (numberOfChildren < sortedDimensionKeys.length) {
				throw `There are fewer children than defined areas for the grid.`;
			}
			// Use cloneElement with an automatically assigned postion: absolute
			// style parameter. This allows positioning attributes of the child
			// element to behave as expected.
			children = React.Children.map(props.children, (child, index) => {
				let dimension =
					(child.props.item && extractedDimensions[child.props.item]) ||
					extractedDimensions[sortedDimensionKeys[index]];
				return (
					<GridCell
						top={dimension.top}
						bottom={dimension.bottom}
						left={dimension.left}
						right={dimension.right}>
						{React.cloneElement(child, {
							style: {...child.style || {}, position: 'absolute'}
						})}
					</GridCell>
				)
			});
		}
		return (
			<GridDiv
				rowHeaders={rowHeaders}
				colHeaders={colHeaders}
				className={props.className}>
				{children}
			</GridDiv>
		);
	};
	// Add properties for unit testing purposes.
	component.rowHeaders = rowHeaders;
	component.colHeaders = colHeaders;
	component.extractedDimensions = extractedDimensions;
	return component;
};
export default GridComponentFactory;