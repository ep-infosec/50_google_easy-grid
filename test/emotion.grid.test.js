import grid from "../emotion";
import React from "react";
import { create as render } from "react-test-renderer";
import serializer, { matchers } from "jest-emotion";
expect.extend(matchers);
expect.addSnapshotSerializer(serializer);

const renderJSON = el => render(el).toJSON();

describe("The grid component should", () => {
  it("accept layout via interpolation", () => {
    const TwoByTwo = grid`
					1fr  2fr
			3fr A,B  A,B
			4fr A    A
		`;
  });

  it("throw on too many header row columns", () => {
    expect(() => {
      grid`
						1fr  2fr
				3fr A,B
				4fr A  
			`;
    }).toThrow(/headers row has too many columns/);
  });

  it("throw on non-uniform column lengths", () => {
    expect(() => {
      grid`
				    1fr
				1fr  A
				1fr
			`;
    }).toThrow(/should have the same number of/);
  });

  it("be able to handle only column headers", () => {
    let Component = null;
    expect(() => {
      Component = grid`1fr 2fr 3fr`;
    }).not.toThrow();
    expect(Component.colHeaders).toEqual("1fr 2fr 3fr");
  });

  it("be able to handle only row headers", () => {
    let Component = null;
    expect(() => {
      Component = grid`
			1fr
			2fr 
			3fr`;
    }).not.toThrow();
    expect(Component.rowHeaders).toEqual("1fr 2fr 3fr");
  });

  it("produce a dimensions map", () => {
    const Component = grid`
						3fr   4fr
				1fr	A,B   A,B
				2fr A     A
			`;
    const areaDimensions = Component.extractedDimensions;
    expect(areaDimensions["A"].top).toBe(0);
    expect(areaDimensions["A"].left).toBe(0);
    expect(areaDimensions["A"].bottom).toBe(1);
    expect(areaDimensions["A"].right).toBe(1);

    expect(areaDimensions["B"].top).toBe(0);
    expect(areaDimensions["B"].left).toBe(0);
    expect(areaDimensions["B"].bottom).toBe(0);
    expect(areaDimensions["B"].right).toBe(1);

    expect(Component.colHeaders).toEqual("3fr 4fr");
    expect(Component.rowHeaders).toEqual("1fr 2fr");
  });

  it("assign dimensions based on sorting", () => {
    const Component = grid`
						3fr   4fr
				1fr	A,B   A,B
				2fr A     A
			`;

    const json = renderJSON(
      <Component>
        <div />
        <div />
      </Component>
    );

    expect(json).toMatchSnapshot();

    const firstGridItem = json.children[0];
    const secondGridItem = json.children[1];

    expect(firstGridItem).toHaveStyleRule("grid-row-start", "1");
    expect(firstGridItem).toHaveStyleRule("grid-row-end", "3");
    expect(firstGridItem).toHaveStyleRule("grid-column-start", "1");
    expect(firstGridItem).toHaveStyleRule("grid-column-end", "3");
    expect(secondGridItem).toHaveStyleRule("grid-row-start", "1");
    expect(secondGridItem).toHaveStyleRule("grid-row-end", "2");
    expect(secondGridItem).toHaveStyleRule("grid-column-start", "1");
    expect(secondGridItem).toHaveStyleRule("grid-column-end", "3");
  });

  it("assign dimensions based on gridItem", () => {
    const Component = grid`
			3fr		4fr 	
		1fr A,B		A,B	
		2fr A		  A	
			`;

    const json = renderJSON(
      <Component>
        <div item="B" />
        <div item="A" />
      </Component>
    );

    expect(json).toMatchSnapshot();
    const firstGridItem = json.children[0];
    const secondGridItem = json.children[1];

    expect(firstGridItem).toHaveStyleRule("grid-row-start", "1");
    expect(firstGridItem).toHaveStyleRule("grid-row-end", "2");
    expect(firstGridItem).toHaveStyleRule("grid-column-start", "1");
    expect(firstGridItem).toHaveStyleRule("grid-column-end", "3");
    expect(secondGridItem).toHaveStyleRule("grid-row-start", "1");
    expect(secondGridItem).toHaveStyleRule("grid-row-end", "3");
    expect(secondGridItem).toHaveStyleRule("grid-column-start", "1");
    expect(secondGridItem).toHaveStyleRule("grid-column-end", "3");
  });

  it("support empty cells", () => {
    const Component = grid`
						3fr   4fr
				1fr	..   ..
				2fr A     A
			`;
    const areaDimensions = Component.extractedDimensions;
    expect(areaDimensions["A"].top).toBe(1);
    expect(areaDimensions["A"].left).toBe(0);
    expect(areaDimensions["A"].bottom).toBe(1);
    expect(areaDimensions["A"].right).toBe(1);

    expect(Component.colHeaders).toEqual("3fr 4fr");
    expect(Component.rowHeaders).toEqual("1fr 2fr");
  });
});
