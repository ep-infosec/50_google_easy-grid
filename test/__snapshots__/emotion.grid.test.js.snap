// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The grid component should assigns dimensions based on gridItem 1`] = `
.emotion-2 {
  display: grid;
  grid-template-rows: 1fr 2fr;
  grid-template-columns: 3fr 4fr;
}

.emotion-1 {
  grid-row-start: 1;
  grid-row-end: 3;
  grid-column-start: 1;
  grid-column-end: 3;
  position: relative;
}

.emotion-0 {
  grid-row-start: 1;
  grid-row-end: 2;
  grid-column-start: 1;
  grid-column-end: 3;
  position: relative;
}

<div
  className="emotion-2"
>
  <div
    className="emotion-0"
  >
    <div
      item="B"
      style={
        Object {
          "position": "absolute",
        }
      }
    />
  </div>
  <div
    className="emotion-1"
  >
    <div
      item="A"
      style={
        Object {
          "position": "absolute",
        }
      }
    />
  </div>
</div>
`;

exports[`The grid component should assigns dimensions based on sorting 1`] = `
.emotion-2 {
  display: grid;
  grid-template-rows: 1fr 2fr;
  grid-template-columns: 3fr 4fr;
}

.emotion-0 {
  grid-row-start: 1;
  grid-row-end: 3;
  grid-column-start: 1;
  grid-column-end: 3;
  position: relative;
}

.emotion-1 {
  grid-row-start: 1;
  grid-row-end: 2;
  grid-column-start: 1;
  grid-column-end: 3;
  position: relative;
}

<div
  className="emotion-2"
>
  <div
    className="emotion-0"
  >
    <div
      style={
        Object {
          "position": "absolute",
        }
      }
    />
  </div>
  <div
    className="emotion-1"
  >
    <div
      style={
        Object {
          "position": "absolute",
        }
      }
    />
  </div>
</div>
`;
